packages/apollo-federation/src/composition/validate/postComposition/keySelectionSetsDeclared.ts

```diff
@@ -0,0 +1,91 @@
+import { GraphQLSchema, isObjectType, FieldNode, GraphQLError } from 'graphql';
+
+import { logServiceAndType, errorWithCode } from '../../utils';
+
+/**
+ * - Each key's selection set must be present on the base service
+ */
+export const keySelectionSetsDeclared = (schema: GraphQLSchema) => {
+  const errors: GraphQLError[] = [];
+
+  const types = schema.getTypeMap();
+  for (const [typeName, namedType] of Object.entries(types)) {
+    if (!isObjectType(namedType)) continue;
+
+    if (namedType.federation && namedType.federation.keys) {
+      const baseService = namedType.federation.serviceName;
+      if (!baseService) {
+        // Federated type does not have a base service. This shouldn't happen, but does in some tests.
+        // Just exit early?
+        continue;
+      }
+
+      const baseKeys = namedType.federation.keys[baseService] as FieldNode[][];
+
+      for (const [serviceName, selectionSets] of Object.entries(
+        namedType.federation.keys,
+      )) {
+        for (const selectionSet of selectionSets as FieldNode[][]) {
+          if (!baseKeys.some(key => selectionSetsEqual(key, selectionSet))) {
+            errors.push(
+              errorWithCode(
+                'KEY_NOT_DECLARED',
+                logServiceAndType(serviceName, typeName) +
+                  `uses the key \`${selectionSetToString(
+                    selectionSet,
+                  )}\`, which is not declared on the base service. All keys must be declared on the base service. `,
+              ),
+            );
+          }
+        }
+      }
+    }
+  }
+
+  return errors;
+};
+
+const selectionSetToString = (a: FieldNode[]): string =>
```

Comment:
Can we use `graphql.print()` for this or is the outcome different from what I'm imagining?

Reply:
It gave weird formatting in my experience, also it doesn't operate that smoothly with the `FieldNode` array, and we don't have a `SelectionSet` to pass it, so idk. Left a comment.